### PR TITLE
Fixes #52 : Discard iterator if the allocator is modified beyond its current position

### LINKED ISSUE #52: OffHeapStorageArea.moveAddressDown() makes dangerous use of the allocator iterator
The `OffHeapStorageArea.moveAddressDown()` method can make mutative changes to the allocator when executing the owners evictAtAddress() that invalidate the 'next' pointer already retrieved by the allocators iterator.  We should look in to reworking this logic to avoid this scenario.

### PR DESCRIPTION
**Pull request by chrisdennis**

### CODE REVIEW

File: src/main/java/org/terracotta/offheapstore/paging/OffHeapStorageArea.java [hunk 1]
```diff
@@ -625,46 +625,53 @@ private boolean moveAddressDown(long target) {
     int sizeOfArea = owner.sizeOf(target);
 
     long ceiling = addressForPage(Math.max(0, pageIndexFor(target) - 2)) + 1;
-    long startAt = random.nextLong() % ceiling; //check for negative results??
+    long startAt = Long.MAX_VALUE & (random.nextInt() % ceiling);
```
Comment (cljohnso) [inline src/main/java/org/terracotta/offheapstore/paging/OffHeapStorageArea.java]: Did you mean to use `nextInt()` here?

File: src/main/java/org/terracotta/offheapstore/paging/OffHeapStorageArea.java [hunk 2]
```diff
@@ -625,46 +625,53 @@ private boolean moveAddressDown(long target) {
...
+                writeBuffers(relocated, readBuffers(target, size));
+                if (!owner.moved(target, relocated)) {
+                  throw new AssertionError("Failure to move mapping during release");
+                }
+                allocator.free(target);
+                return true;
+              } else {
+                allocator.free(relocated);
+
+                if (removed.stream().anyMatch(p -> p > address)) {
```
Comment (cljohnso) [inline src/main/java/org/terracotta/offheapstore/paging/OffHeapStorageArea.java]: I have questions about this ...

File: src/main/java/org/terracotta/offheapstore/paging/OffHeapStorageArea.java (hunk 1 above)
Comment (chrisdennis) [reply src/main/java/org/terracotta/offheapstore/paging/OffHeapStorageArea.java]: No this should still be `nextLong()`

**Comment by chrisdennis** [comment]
Replacing with daggy PR.

**Pull request closed without merging**